### ISSUE TITLE
Lime image explainer with yolov3 keras (or darknet)

### ISSUE THREAD

**Issue by wyzrxrs** [marcotcr/lime#488]
Hello,
I would like to use Lime to explain the predictions from the yolov3 keras model. I have issues on the following explanation line.

explanation = explainer.explain_instance(image_to_explain, predict_function, hide_color=0, top_labels=5,  num_samples=1000)

My first question:  the image_to_explain as the first parameter, what is the shape of this image, is it 3-dimension or 4-dimension? 

Second, for the predict_function, the pretrained model 'model.h5' does not seem to provide a predict function directly.  Is there a way to produce the output array with prediction of all labels and all probabilities ?

Do you have a sample use case for yolov3 keras? It could also be examples using Lime with models from Yolov3 Darknet, I have the similar issues regarding the predict function. So, having examples would be of great help.

Thanks!

**Comment by marcotcr** [comment]
[image](https://github.com/marcotcr/lime/blob/master/lime/lime_image.py#L146): 3 dimension RGB image. If this is only two dimensional, we will assume it's a grayscale image and call gray2rgb.

> Second, for the predict_function, the pretrained model 'model.h5' does not seem to provide a predict function directly. Is there a way to produce the output array with prediction of all labels and all probabilities ?

I haven't used yolov3, but if you don't have a prediction function how do you use the model? : )

**Comment by wyzrxrs** [comment]
Thanks! Yolov3 requires a 4d-image, but the explainer requires an input to be a 3d-image. So the following solves the problem.
np.expand_dims(img_3d, axis=0)

For the predict function, yes, there exists a prediction function from yolo3, although the output format is not the same as what's expected by lime. So finally I found a way around, using a wrapper function. 

In case someone gets stuck, this is a sample code for the predict function, using yolov3 trained model.
`
def predict_func(ls_img):
    predictions = []
    for img_3d in ls_img:
        img_4d = transform_func(img_3d)    
        yhat = model.predict(img_4d)
    ...
       v_boxes, v_labels, v_scores = get_boxes(boxes, labels, class_threshold)
    ...
       temp_dict = {}
        # summarize what we found
        for i in range(len(v_boxes)):
            if v_labels[i] not in temp_dict.keys():
                temp_dict[v_labels[i]] = v_scores[i]
                out.append((v_labels[i], v_scores[i]))
        predictions.append(np.array([temp_dict.get(label, 0.0) for label in labels]))
       return np.array(predictions)
`

**Comment by AntMorais** [comment]
Hi @wyzrxrs , I'm having the same problem as you had, but in my case I want to get a predict function for YOLOv4 darknet (AlexeyAB's repository). Any idea on how to do the same thing you did in the previous comment but in darknet? Thanks!

**Comment by wyzrxrs** [comment]
Hello, sorry I haven't worked on the darknet version you've used. There must exist a predict function, not sure, could look into this function detect_image() , or something else. It would be great to share your findings afterwards. Thanks!

**Comment by AntMorais** [comment]
Here it is on Darknet YOLOv4 https://github.com/AntMorais/yolime. I'll add a README when I can.

**Comment by nn09** [comment]
@AntMorais Were you able to solve the issue and call LIME on YOLO darknet models?

**Comment by AntMorais** [comment]
Yes, you can find it [here](https://github.com/AntMorais/yolime).

**Comment by m10913008** [comment]
@AntMorais Excuse me, in your example(https://github.com/AntMorais/yolime), how to generate probabilityArray.txt file,It looks like probabilities for each category item, but I don't know how to get these data

**Comment by AntMorais** [comment]
In [this function](https://github.com/AntMorais/yolime/blob/d5f68cdd0f5b59dbc3cdcacc973833aa2799096d/src/yolo_layer.c#L861) you can see where I get the class probabilities (line 924).
You could do a `git diff` between my repo and the parent repo to see what I changed if that helps :)

**Comment by m10913008** [comment]
@AntMorais Thank you for your reply, it helped me a lot
